File: infrastructures/rest-ddd/typescript/src/application/mappers/ExampleMapper.ts
/**
 * Example Mapper
 * MANDATORY for all transformations between layers.
 */

import { IMapper } from './IMapper';
import { ExampleModel, ExampleStatus } from '../../domain/models/ExampleModel';
import { ExampleWriteEntity } from '../../infrastructure/persistence/write/ExampleWriteEntity';
import { ExampleReadEntity } from '../../infrastructure/persistence/read/ExampleReadEntity';
import { CreateExampleRequest } from '../../api/dto/CreateExampleRequest';
import { ExampleResponse } from '../../api/dto/ExampleResponse';

export class ExampleMapper implements IMapper<
  CreateExampleRequest,
  ExampleModel,
  ExampleWriteEntity,
  ExampleReadEntity,
  ExampleResponse
> {
  /**
   * Mapper for Example entity.
   * 
   * Rules:
   * - Located in /application/mappers/
   * - Handles all transformations: DTO ↔ BMO ↔ Entity
   * - MANDATORY for all layer transitions
   */

  toModelFromRequest(p_dto: CreateExampleRequest): ExampleModel {
    return new ExampleModel(
      p_dto.name,
      p_dto.description,
      ExampleStatus.PENDING,
      p_dto.ownerId
    );
  }

  toWriteEntity(p_model: ExampleModel): ExampleWriteEntity {
    return new ExampleWriteEntity(
      p_model.name,
      p_model.description,
      p_model.status,
      p_model.ownerId,
      p_model.id
    );
  }

  toModelFromWriteEntity(p_entity: ExampleWriteEntity): ExampleModel {
    return new ExampleModel(
      p_entity.name,
      p_entity.description,
      p_entity.status as ExampleStatus,
      p_entity.ownerId,
      p_entity.id
    );
  }

  toResponseFromReadEntity(p_entity: ExampleReadEntity): ExampleResponse {
    return {
      id: p_entity.id!,
      name: p_entity.name,
      description: p_entity.description,
      status: p_entity.status,
      ownerId: p_entity.ownerId,
      ownerName: p_entity.ownerName,
      createdAt: p_entity.createdAt,
      updatedAt: p_entity.updatedAt
    };
  }

  toResponseFromModel(p_model: ExampleModel): ExampleResponse {
    return {
      id: p_model.id!,
      name: p_model.name,
      description: p_model.description,
      status: p_model.status,
      ownerId: p_model.ownerId,
      ownerName: '', // Not available in model
      createdAt: new Date().toISOString(),
      updatedAt: new Date().toISOString()
    };
  }
}

File: infrastructures/rest-ddd/csharp/src/Application/Mappers/ExampleMapper.cs
/**
 * Example Mapper
 * MANDATORY for all transformations between layers.
 */

using Api.Dto;
using Domain.Models;
using Infrastructure.Persistence.Read;
using Infrastructure.Persistence.Write;
using System;

namespace Application.Mappers
{
    public class ExampleMapper : IMapper<
        CreateExampleRequest,
        ExampleModel,
        ExampleWriteEntity,
        ExampleReadEntity,
        ExampleResponse
    >
    {
        /**
         * Mapper for Example entity.
         * 
         * Rules:
         * - Located in /Application/Mappers/
         * - Handles all transformations: DTO ↔ BMO ↔ Entity
         * - MANDATORY for all layer transitions
         */

        public ExampleModel ToModelFromRequest(CreateExampleRequest p_dto)
        {
            return new ExampleModel(
                p_dto.Name,
                p_dto.Description,
                ExampleStatus.Pending,
                p_dto.OwnerId
            );
        }

        public ExampleWriteEntity ToWriteEntity(ExampleModel p_model)
        {
            return new ExampleWriteEntity(
                p_model.Name,
                p_model.Description,
                p_model.Status.ToString(),
                p_model.OwnerId
            )
            {
                Id = p_model.Id
            };
        }

        public ExampleModel ToModelFromWriteEntity(ExampleWriteEntity p_entity)
        {
            return new ExampleModel(
                p_entity.Name,
                p_entity.Description,
                Enum.Parse<ExampleStatus>(p_entity.Status),
                p_entity.OwnerId,
                p_entity.Id
            );
        }

        public ExampleResponse ToResponseFromReadEntity(ExampleReadEntity p_entity)
        {
            return new ExampleResponse
            {
                Id = p_entity.Id!,
                Name = p_entity.Name,
                Description = p_entity.Description,
                Status = p_entity.Status,
                OwnerId = p_entity.OwnerId,
                OwnerName = p_entity.OwnerName,
                CreatedAt = p_entity.CreatedAt,
                UpdatedAt = p_entity.UpdatedAt
            };
        }

        public ExampleResponse ToResponseFromModel(ExampleModel p_model)
        {
            return new ExampleResponse
            {
                Id = p_model.Id!,
                Name = p_model.Name,
                Description = p_model.Description,
                Status = p_model.Status.ToString(),
                OwnerId = p_model.OwnerId,
                OwnerName = string.Empty,  // Not available in model
                CreatedAt = DateTime.UtcNow.ToString("o"),
                UpdatedAt = DateTime.UtcNow.ToString("o")
            };
        }
    }
}
